pkg/proc/native/proc_freebsd.go

```diff
@@ -0,0 +1,368 @@
+package native
+
+// #cgo LDFLAGS: -lprocstat
+// #include <stdlib.h>
+// #include "proc_freebsd.h"
+import "C"
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	sys "golang.org/x/sys/unix"
+
+	"github.com/go-delve/delve/pkg/proc"
+
+	isatty "github.com/mattn/go-isatty"
+)
+
+// Process statuses
+const (
+	StatusIdle     = 1
+	StatusRunning  = 2
+	StatusSleeping = 3
+	StatusStopped  = 4
+	StatusZombie   = 5
+	StatusWaiting  = 6
+	StatusLocked   = 7
+)
+
+// OSProcessDetails contains FreeBSD specific
+// process details.
+type OSProcessDetails struct {
+	comm string
```

Comment:
I can't find where this is used.

Reply:
Seems it's not used anywhere. I've tried to use tid (main process thread id) in singleStep mimicking comparison in line 59 from linux_threads.go:
`if (status == nil || status.Exited()) && wpid == t.dbp.pid`
But later replaced all logic in singleStep with a single call to trapWait(). Btw can we do the same for Linux? The problem with Linux's singleStep implementation is that it doesn't handle threads add\remove properly. I see that we are waiting for a TRAP from "this" thread ignoring signals from other threads. Maybe we can just pass tid into trapWait as additional param to simulate the same?

Reply:
> Btw can we do the same for Linux?

see the comment in (*Process).wait in proc_linux.go

Reply:
The only comment there is about a workaround for lock when main thread exits and leaves zombies. Could you clarify how does that rely to replacing logic in singleStep() with a call to waitTrap()?

Reply:
The full wait is slow because of the workaround. Also all threads are stopped and we are only executing one instruction of threads stopped on a breakpoint, so it's fine.

Reply:
On FreeBSD, even when all threads are stopped, wait in cycle may return multiple queued signals from multiple stopped threads.
 
I'm preparing a detailed step by step explanation of why "TestBreakpointCounts" fails, hopefully someone can help with a fix.